File: src/installer.py
"""Install / update / uninstall plugins by manipulating ~/.claude/plugins/.

A marketplace is just an *index*. The actual plugin lives at the location given
by its `source` entry in `marketplace.json`. So install means:

 1. Resolve the plugin's source (a GitHub URL or directory path).
 2. Download/copy the plugin source at the configured ref into
    `~/.claude/plugins/cache/<marketplace>/<plugin>/<version>/`.
 3. Patch `installed_plugins.json` with a fresh install record.
"""
from __future__ import annotations

import json
import os
import shutil
import stat
from datetime import datetime, timezone
from pathlib import Path

from . import config, plugin_state
from .github_client import GitHubClient
from .models import Plugin, PluginSource


def _rmtree_robust(path: Path) -> None:
    """Remove a directory tree even when paths exceed Windows MAX_PATH or files are read-only."""
    if not path.exists():
        return

    def _onerror(func, p, _exc):
        try:
            os.chmod(p, stat.S_IWRITE)
            func(p)
        except Exception:
            pass

    target = str(path)
    if os.name == "nt" and not target.startswith("\\\\?\\"):
        target = "\\\\?\\" + str(path.resolve())
    shutil.rmtree(target, onerror=_onerror)


def _atomic_write_json(path: Path, data: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(data, indent=2), encoding="utf-8")
    tmp.replace(path)


def _load_installed() -> dict:
    f = config.installed_plugins_file()
    if not f.exists():
        return {"version": 2, "plugins": {}}
    try:
        return json.loads(f.read_text(encoding="utf-8"))
    except Exception:
        return {"version": 2, "plugins": {}}


def _save_installed(data: dict) -> None:
    _atomic_write_json(config.installed_plugins_file(), data)


def _now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def _cache_path(marketplace: str, plugin: str, version: str) -> Path:
    return config.plugins_cache_dir() / marketplace / plugin / version


def install_plugin(gh: GitHubClient, plugin: Plugin) -> Path:
    """Install (or update) a plugin using its embedded `source` info.

    Raises ValueError if the plugin has no resolvable source.
    """
    if not plugin.source:
        raise ValueError(f"Plugin {plugin.name} has no source — cannot install.")
    src: PluginSource = plugin.source
    version = plugin.latest_version or src.ref or "0.0.0"

    if src.repo:
        ref = src.ref or gh.get_default_branch(src.repo)
        return _install_from_github(gh, src.repo, plugin.marketplace_name, plugin.name, version, ref)
    if src.kind == "directory" and src.path:
        return _install_from_directory(Path(src.path), plugin.marketplace_name, plugin.name, version)
    raise ValueError(f"Plugin {plugin.name}: unsupported source ({src.kind}: {src.url or src.path})")


def _install_from_github(gh: GitHubClient, repo: str, marketplace_name: str,
                         plugin_name: str, version: str, ref: str) -> Path:
    zip_bytes = gh.download_zipball(repo, ref=ref)
    install_path = _cache_path(marketplace_name, plugin_name, version)
    _rmtree_robust(install_path)
    install_path.mkdir(parents=True, exist_ok=True)
    GitHubClient.extract_zipball(zip_bytes, install_path)

    sha = ""
    try:
        commit = gh.get_latest_commit(repo, branch=ref)
        sha = commit.get("sha", "")
    except Exception:
        sha = ref if len(ref) == 40 else ""

    _register_install(marketplace_name, plugin_name, version, install_path, sha)
    return install_path


def _install_from_directory(source_dir: Path, marketplace_name: str,
                            plugin_name: str, version: str) -> Path:
    install_path = _cache_path(marketplace_name, plugin_name, version)
    _rmtree_robust(install_path)
    shutil.copytree(source_dir, install_path)
    _register_install(marketplace_name, plugin_name, version, install_path, "")
    return install_path


def _register_install(marketplace_name: str, plugin_name: str, version: str,
                      install_path: Path, git_sha: str) -> None:
    data = _load_installed()
    plugins = data.setdefault("plugins", {})
    key = f"{plugin_name}@{marketplace_name}"
    record = {
        "scope": "user",
        "installPath": str(install_path),
        "version": version,
        "installedAt": _now_iso(),
        "lastUpdated": _now_iso(),
    }
    if git_sha:
        record["gitCommitSha"] = git_sha
    plugins[key] = [record]
    data["version"] = data.get("version", 2)
    _save_installed(data)
    # Mirror Claude Code's `/plugin install`: newly installed plugins are enabled
    # by default, but only if the user hasn't already set an explicit value.
    if plugin_state.get_enabled(plugin_name, marketplace_name) is None:
        plugin_state.set_enabled(plugin_name, marketplace_name, True)


def uninstall(plugin: Plugin) -> None:
    data = _load_installed()
    key = f"{plugin.name}@{plugin.marketplace_name}"
    if key in data.get("plugins", {}):
        data["plugins"].pop(key, None)
        _save_installed(data)
    if plugin.install_path and plugin.install_path.exists():
        _rmtree_robust(plugin.install_path)
    plugin_state.remove_entry(plugin.name, plugin.marketplace_name)

File: src-tauri/src/installer.rs
//! Install / update / uninstall plugins — port of src/installer.py.

use crate::config;
use crate::error::{Error, Result};
use crate::github_client::{long_path, GitHubClient};
use crate::models::Plugin;
use crate::plugin_state;
use chrono::Utc;
use serde_json::{json, Map, Value};
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};

/// Remove a directory tree even when paths exceed Windows MAX_PATH or files are read-only.
pub fn rmtree_robust(path: &Path) -> Result<()> {
    if !path.exists() {
        return Ok(());
    }
    let target = long_path(path);
    // Walk and chmod read-only files before letting remove_dir_all do its job.
    for entry in walkdir::WalkDir::new(&target)
        .contents_first(true)
        .into_iter()
        .flatten()
    {
        let p = entry.path();
        if let Ok(meta) = p.metadata() {
            let mut perms = meta.permissions();
            #[allow(clippy::permissions_set_readonly_false)]
            perms.set_readonly(false);
            let _ = fs::set_permissions(p, perms);
        }
    }
    fs::remove_dir_all(&target).map_err(Error::from)?;
    Ok(())
}

pub fn atomic_write_json(path: &Path, data: &Value) -> Result<()> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent)?;
    }
    let tmp = path.with_extension(format!(
        "{}.tmp",
        path.extension().and_then(|e| e.to_str()).unwrap_or("json")
    ));
    let mut f = OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .open(&tmp)?;
    f.write_all(serde_json::to_string_pretty(data)?.as_bytes())?;
    drop(f);
    fs::rename(&tmp, path)?;
    Ok(())
}

fn load_installed() -> Value {
    let f = config::installed_plugins_file();
    if !f.exists() {
        return json!({"version": 2, "plugins": {}});
    }
    fs::read_to_string(&f)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_else(|| json!({"version": 2, "plugins": {}}))
}

fn save_installed(data: &Value) -> Result<()> {
    atomic_write_json(&config::installed_plugins_file(), data)
}

pub fn now_iso() -> String {
    let now = Utc::now();
    // Match Python: %Y-%m-%dT%H:%M:%S.<ms>Z
    now.format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string()
}

fn cache_path(marketplace: &str, plugin: &str, version: &str) -> PathBuf {
    config::plugins_cache_dir().join(marketplace).join(plugin).join(version)
}

pub fn install_plugin(gh: &GitHubClient, plugin: &Plugin) -> Result<PathBuf> {
    tracing::info!(
        "install_plugin: {} from {} (marketplace={})",
        plugin.name,
        plugin
            .source
            .as_ref()
            .map(|s| s.repo.as_str())
            .unwrap_or("?"),
        plugin.marketplace_name
    );
    let src = plugin
        .source
        .as_ref()
        .ok_or_else(|| {
            tracing::error!("plugin {} has no source", plugin.name);
            Error::Invalid(format!("Plugin {} has no source.", plugin.name))
        })?;
    let version = plugin
        .latest_version
        .clone()
        .or_else(|| Some(src.r#ref.clone()))
        .filter(|v| !v.is_empty())
        .unwrap_or_else(|| "0.0.0".to_string());

    if !src.repo.is_empty() {
        let r#ref = if src.r#ref.is_empty() {
            gh.get_default_branch(&src.repo)?
        } else {
            src.r#ref.clone()
        };
        return install_from_github(
            gh,
            &src.repo,
            &plugin.marketplace_name,
            &plugin.name,
            &version,
            &r#ref,
        );
    }
    if src.kind == "directory" && !src.path.is_empty() {
        return install_from_directory(
            Path::new(&src.path),
            &plugin.marketplace_name,
            &plugin.name,
            &version,
        );
    }
    Err(Error::Invalid(format!(
        "Plugin {}: unsupported source ({}: {})",
        plugin.name,
        src.kind,
        if !src.url.is_empty() { &src.url } else { &src.path }
    )))
}

fn install_from_github(
    gh: &GitHubClient,
    repo: &str,
    marketplace_name: &str,
    plugin_name: &str,
    version: &str,
    r#ref: &str,
) -> Result<PathBuf> {
    let zip_bytes = gh.download_zipball(repo, r#ref)?;
    let install_path = cache_path(marketplace_name, plugin_name, version);
    rmtree_robust(&install_path)?;
    fs::create_dir_all(&install_path)?;
    GitHubClient::extract_zipball(&zip_bytes, &install_path, "")?;

    let sha = gh
        .get_latest_commit(repo, r#ref)
        .ok()
        .and_then(|v| v.get("sha").and_then(|s| s.as_str()).map(String::from))
        .unwrap_or_else(|| {
            if r#ref.len() == 40 {
                r#ref.to_string()
            } else {
                String::new()
            }
        });

    register_install(marketplace_name, plugin_name, version, &install_path, &sha)?;
    Ok(install_path)
}

fn install_from_directory(
    source_dir: &Path,
    marketplace_name: &str,
    plugin_name: &str,
    version: &str,
) -> Result<PathBuf> {
    let install_path = cache_path(marketplace_name, plugin_name, version);
    rmtree_robust(&install_path)?;
    copy_dir_all(source_dir, &install_path)?;
    register_install(marketplace_name, plugin_name, version, &install_path, "")?;
    Ok(install_path)
}

fn copy_dir_all(src: &Path, dst: &Path) -> Result<()> {
    fs::create_dir_all(dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let target = dst.join(entry.file_name());
        if entry.file_type()?.is_dir() {
            copy_dir_all(&entry.path(), &target)?;
        } else {
            fs::copy(entry.path(), &target)?;
        }
    }
    Ok(())
}

fn register_install(
    marketplace_name: &str,
    plugin_name: &str,
    version: &str,
    install_path: &Path,
    git_sha: &str,
) -> Result<()> {
    let mut data = load_installed();
    let plugins = data
        .as_object_mut()
        .unwrap()
        .entry("plugins".to_string())
        .or_insert_with(|| Value::Object(Map::new()));
    let plugins_obj = plugins.as_object_mut().unwrap();
    let key = format!("{plugin_name}@{marketplace_name}");
    let mut record = json!({
        "scope": "user",
        "installPath": install_path.to_string_lossy(),
        "version": version,
        "installedAt": now_iso(),
        "lastUpdated": now_iso(),
    });
    if !git_sha.is_empty() {
        record["gitCommitSha"] = json!(git_sha);
    }
    plugins_obj.insert(key, json!([record]));
    let obj = data.as_object_mut().unwrap();
    if !obj.contains_key("version") {
        obj.insert("version".to_string(), json!(2));
    }
    save_installed(&data)?;
    if plugin_state::get_enabled(plugin_name, marketplace_name).is_none() {
        plugin_state::set_enabled(plugin_name, marketplace_name, true)?;
    }
    Ok(())
}

pub fn uninstall(plugin: &Plugin) -> Result<()> {
    tracing::info!(
        "uninstall_plugin: {} (marketplace={})",
        plugin.name,
        plugin.marketplace_name
    );
    let mut data = load_installed();
    let key = format!("{}@{}", plugin.name, plugin.marketplace_name);
    let mut changed = false;
    if let Some(plugins) = data.as_object_mut().and_then(|d| d.get_mut("plugins")) {
        if let Some(obj) = plugins.as_object_mut() {
            if obj.remove(&key).is_some() {
                changed = true;
            }
        }
    }
    if changed {
        save_installed(&data)?;
    }
    if let Some(p) = &plugin.install_path {
        if p.exists() {
            rmtree_robust(p)?;
        }
    }
    plugin_state::remove_entry(&plugin.name, &plugin.marketplace_name)?;
    Ok(())
}

